grakel_replace/optimize.py

```diff
@@ -1,14 +1,49 @@
 from __future__ import annotations
 
 import random
+from collections.abc import Iterator
+from contextlib import contextmanager
 from typing import TYPE_CHECKING
 
 import networkx as nx
 import torch
 from botorch.optim import optimize_acqf_mixed
+from grakel_replace.torch_wl_kernel import TorchWLKernel
 
 if TYPE_CHECKING:
     from botorch.acquisition import AcquisitionFunction
+    from botorch.models.gp_regression_mixed import Kernel
+
+
+# Making predictions on test data
+# No the wl_kernel needs to be aware of the test graphs
+@contextmanager
+def set_graph_lookup(
+    kernel: Kernel,
+    new_graphs: list[nx.Graph],
+    *,
+    append: bool = True,
+) -> Iterator[None]:
+    kernel_prev_graphs: list[tuple[Kernel, list[nx.Graph]]] = []
+    if isinstance(kernel, TorchWLKernel):
+        modules = [kernel]
+    else:
+        assert hasattr(
+            kernel, "sub_kernels"
+        ), "Kernel module must have sub_kernels method."
+        modules = [k for k in kernel.sub_kernels() if isinstance(k, TorchWLKernel)]
+
+    for kern in modules:
+        kernel_prev_graphs.append((kern, kern.graph_lookup))
+        if append:
+            kern.set_graph_lookup([*kern.graph_lookup, *new_graphs])
+        else:
+            kern.set_graph_lookup(new_graphs)
+
+    yield
+
+    for _kern, _prev_graphs in kernel_prev_graphs:
+        _kern.set_graph_lookup(_prev_graphs)
```

Comment:
This is probably the version of the context manager we want. I was just lazy and copy pasted it here for now. Basically the same as the other one